libp2p/8-implementations.md

```diff
@@ -55,3 +55,68 @@ Current known implementations (or WIP) are:
   - Python - <https://github.com/candeira/py-ipfs/blob/readme-roadmap/README.md>
   - Rust - <https://github.com/diasdavid/rust-libp2p>
 
+## 8.1 Swarm
+
+### 8.1.1 Swarm Dialer
+
+The swarm dialer manages making a successful connection to a target peer, given a stream of addresses as inputs, and making sure to respect any and all rate limits imposed. To this end, we have designed the following logic for dialing:
+
+```
+DialPeer(peerID) {
+	if PeerIsBeingDialed(peerID) {
+		waitForDialToComplete(peerID)
+		return BestConnToPeer(peerID)
+	}
+	
+	StartDial(peerID)
+
+	waitForDialToComplete(peerID)
+	return BestConnToPeer(peerID)
+}
+
+	
+StartDial(peerID) {
+	addrs = getAddressStream(peerID)
+
+	addrs.onNewAddr(function(addr) {
+		if rateLimitCanDial(peerID, addr) {
+			doDialAsync(peerID, addr)
+		} else {
+			rateLimitScheduleDial(peerID, addr)
+		}
+	})
+}
+
+// doDialAsync starts dialing to a specific address without blocking.
+// when the dial returns, it releases rate limit tokens, and if it
+// succeeded, will finalize the dial process.
+doDialAsync(peerID, addr) {
```

Comment:
One 'bug' i just caught in writing more tests for this is that we shouldnt start dials if they have been cancelled. For example, I queue up 100 dials, and one in the first ten or so completes successfully we don't want to continue dialing the remaining 90 queued addresses for that peer.

cc @diasdavid